README
NAME

    WWW::FreeProxyListsCom - get proxy lists from
    http://www.freeproxylists.com

SYNOPSIS

        use strict;
        use warnings;
    
        use WWW::FreeProxyListsCom;
    
        my $prox = WWW::FreeProxyListsCom->new;
    
        my $ref = $prox->get_list( type => 'non_anonymous' )
            or die $prox->error;
    
        print "Got a list of " . @$ref . " proxies\nFiltering...\n";
    
        $ref = $prox->filter( port => qr/(80){1,2}/ );
    
        print "Filtered list contains: " . @$ref . " proxies\n"
                . join "\n", map( "$_->{ip}:$_->{port}", @$ref), '';

DESCRIPTION

    The module provides interface to fetch proxy server lists from
    http://www.freeproxylists.com/

CONSTRUCTOR

 new

        my $prox = WWW::FreeProxyListCom->new;
    
        my $prox2 = WWW::FreeProxyListCom->new(
            timeout     => 20, # or 'mech'
            mech        => WWW::Mechanize->new( agent => 'foos', timeout => 20 ),
            debug       => 1,
        );

    Bakes up and returns a fresh WWW::FreeProxyListCom object. Takes a few
    arguments, all of which are optional. Possible arguments are as
    follows:

  timeout

        my $prox = WWW::FreeProxyListCom->new( timeout => 10 );

    Takes a scalar as a value which is the value that will be passed to the
    WWW::Mechanize object to indicate connection timeout in seconds.
    Defaults to: 30 seconds

  mech

        my $prox = WWW::FreeProxyListCom->new(
            mech => WWW::Mechanize->new( agent => '007', timeout => 10 ),
        );

    If a simple timeout is not enough for your needs feel free to specify
    the mech argument which takes a WWW::Mechanize object as a value.
    Defaults to: plain WWW::Mechanize object with timeout argument set to
    whatever WWW::FreeProxyListCom's timeout argument is set to as well as
    agent argument is set to mimic FireFox.

  debug

        my $prox = WWW::FreeProxyListCom->new( debug => 1 );

    When set to a true value will make the object print out some debugging
    info. Defaults to: 0

METHODS

 get_list

        my $list_ref = $prox->get_list
            or die $prox->error;
    
        my $list_ref2 = $prox->get_list(
            type        => 'standard',
            max_pages   => 5,
        ) or die $prox->error;

    Instructs the object ot fetch a list of proxies from
    http://www.freeproxylists.com/ website. On failure returns either undef
    or an empty list depending on the context and the reason for failure
    will be available via error() method. Note: if request for a each of
    the "list" (see max_pages argument below) fails the get_list() will NOT
    error out, if you are getting empty proxy lists try setting debug
    option on in the constructor and it will carp() any failures on the
    "list" gets. On success returns an arrayref of hashrefs, see RETURN
    VALUE section below for details. Takes several arguments all of which
    are optional. To understand them better you should visit
    http://www.freeproxylists.com/ first. The possible arguments are as
    follows:

  type

        ->get_list( type => 'standard' );

    Optional. Specifies the list of proxies to fetch. Defaults to: elite.
    Possible arguments are as follows (valid type values are on the left,
    corresponding "list" site's menu link names are on the right):

        elite           => http elite proxies
        anonymous       => http anonymous lists
        non_anonymous   => http non-anonymous
        https           => https (SSL enabled)
        standard        => http standard ports
        us              => us proxies only
        socks           => socks (version 4/5)

  max_pages

        ->get_list( max_pages => 4 );

    Optional. Specifies how many "lists" to fetch. In other words, if you
    go to list section titled "http elite proxies" you'll see several lists
    in the table; the max_pages specifies how many of those lists to fetch.
    If max_pages is larger than the number of available lists only the
    number of available lists will be fetched. A special value of 0
    indicates that the object should fetch all available lists for a
    specified type. Defaults to: 1 (which is more than enough).

  RETURN VALUE

        $VAR1 = [
            {
                'country' => 'China',
                'last_test' => '3/15 4:23:14 pm',
                'ip' => '121.15.200.147',
                'latency' => '5115',
                'port' => '80',
                'is_https' => 'true'
            },
        ]

    On success get_list() method returns a (possibly empty) arrayref of
    "proxy" hashrefs. The hashrefs represent each proxy listed on the proxy
    list on the site. Each will contain the following keys (if the value
    for a specific key was not found on the site it will be set to N/A):

    ip

      The IP address of the proxy

    port

      The port of the proxy

    country

      The country of the proxy

    last_test

      When was the proxy last tested to be alive, this is the "Date
      checked, UTC" column on the site.

    latency

      Corresponds to the "Latency" column on the site

    is_https

      Corresponds to "HTTPS" column on the site.

 filter

        my $filtered_list_ref = $prox->filter(
            port        => 80,
            ip          => qr/^120/,
            country     => 'Russia',
            is_https    => 'true',
            last_test   => qr|^3/15|, # march 15's
            latency     => qr/\d{1,2}/,
        );

    Must be called after a successfull call to get_list() will croak
    otherwise. Takes one or more key/value pairs of arguments which specify
    filtering rule. The keys are the same as the keys of "proxy" hashref in
    the return value of the get_list() method. Values can be either simple
    scalars or regexes (qr//). If value is a regex the corresponding value
    in the "proxy" hashref will matched against the regex, otherwise the eq
    will be done. Returns an arrayref of "proxy" hashrefs in the exact same
    format as get_list() returns except filtered. In other words calling
    $prox->filter( port => 80, latency => qr/\d{1,2}/ ) will return only
    proxies with port number 80 and for which latency is a two digit value.
    On failure returns either undef or an empty list depending on the
    context and the reason for the error will be available via error()
    method. Although, filter() should not fail if you pass proper filter
    arguments and call it after successfull get_list().

 error

        my $list_ref = $prox->get_list
            or die $prox->error;

    When either get_list() or filter() methods fail they will return either
    undef or an empty list depending on the context and the reason for the
    failure will be available via error() method. Takes no arguments,
    returns a human parsable message explaining why get_list() or filter()
    failed.

 list

        my $last_list_ref = $prox->list;

    Must be called after a successfull call to get_list(). Takes no
    arugments, returns the same arrayref of hashrefs last call to
    get_list() returned.

 filtered_list

        my $last_filtered_list_ref = $prox->filtered_list;

    Must be called after a successfull call to filter(). Takes no
    arugments, returns the same arrayref of hashrefs last call to filter()
    returned.

 mech

        my $old_mech = $prox->mech;
    
        $prox->mech( WWW::Mechanize->new( agent => 'blah' ) );

    Returns a WWW::Mechanize object used for fetching proxy lists. When
    called with an optional argument (which must be a WWW::Mechanize
    object) will use it in any subsequent get_list() calls.

 debug

        my $old_debug = $prox->debug;
    
        $prox->debug( 1 );

    Returns a currently set debug flag (see debug argument to constructor).
    When called with an argument will set the debug flag to the value
    specified.

AUTHOR

    Steve Bertrand <steveb at cpan.org>

    Adopted on Feb 4, 2016 from:

    Zoffix Znet, <zoffix at cpan.org> (http://zoffix.com,
    http://haslayout.net)

BUGS

    Please report any bugs or feature requests to
    https://github.com/stevieb9/p5-www-freeproxylistscom/issues.

COPYRIGHT & LICENSE

    Copyright 2016 Steve Bertrand

    Copyright 2008 Zoffix Znet, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for WWW-FreeProxyListsCom

1.007 UNREL
        - remove perl 5.8 from Travis-CI

1.006   2018-02-07
        - added urls(), returns a list of stringified proxy URLs. Perfect for
          using with WWW::ProxyChecker
        - added t/07 to test the new urls() method
        - modified naming conventions in POD of certain variables (didn't
          modify structure or content)

1.005   2016-02-19
        - removed t/kwalitee.t

1.004   2016-02-16
        - changed use() to require() in POD test eval()s
        - fixed typo in regex check in t/05

1.003   2016-02-08
        - added Test::Perl::Critic tests
        - get_list() in t/00-load.t wasn't passing in the type param, which
          resulted in the default 'elite' type being used on each test
          iteration (fixes #1)
        - moved xt/ tests to t/
        - corrected code that was breaking new critic tests (fixes #2)
        - added _url() method which sets/fetches 'url' param for testing
        - added t/03-error test
        - changed tests that fail in t/02-fetch to properly skip
        - added t/04-filter tests
        - added t/05-timeout tests (closes #4; as invalid)
        - reduced timeout in t/04-filter, and wrapped in eval (closes #5)
        - added tests for _set_error() (closes #6)

1.002   2016-02-07
        - Adopted by Steve Bertrand (STEVEB) (2016-02-04)
        - added git info to Makefile.PL
        - added .travis.yml
        - updated license, removed LICENSE file
        - added travis/coverage badges

1.001   2013-12-27
        - Fixed broken test and prereqs

0.004   2013-12-26
        - Fixed a partially broken test
        - Minor fixes to comply with the specs and Kwalitee game
        - Added autogeneration of README from POD
        - Moved author tests to xt/
        - Switched away from now deprecated Class::Data::Accessor
          to Class::Accessor::Grouped

0.003   2010-08-06
        - Fixed first element of the list being garbage

0.002   2008-03-16
        - Fixed improper prerequisites list which was missing
          Class::Data::Accessor

0.001   2008-03-15
        - First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'WWW::FreeProxyListsCom',
    license             => 'perl',
    dist_author         => 'Zoffix Znet <zoffix@cpan.org>',
    dist_version_from   => 'lib/WWW/FreeProxyListsCom.pm',
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'Carp'                     => 1.04,
        'URI'                      => 1.35,
        'WWW::Mechanize'           => 1.34,
        'HTML::TokeParser::Simple' => 3.15,
        'HTML::Entities'           => 1.35,
        'Devel::TakeHashArgs'      => 0.001,
        'Class::Accessor::Grouped'    => 0,
        'perl' => '5.006',
    },
    configure_requires => { 'Module::Build' => 0 },
    add_to_cleanup      => [ 'WWW-FreeProxyListsCom-*' ],
    create_makefile_pl => 'traditional',
);
if ( $ENV{RELEASE_TESTING} ) {
    system('pod2text lib/WWW/FreeProxyListsCom.pm > README');
}
$builder->create_build_script();


Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4007
require 5.006;
use ExtUtils::MakeMaker;
WriteMakefile
(
    'NAME' => 'WWW::FreeProxyListsCom',
    'VERSION_FROM' => 'lib/WWW/FreeProxyListsCom.pm',
    'META_MERGE' => {
    'meta-spec' => { version => 2 },
        resources => {
            bugtracker => {
                web => 'https://github.com/stevieb9/p5-www-freeproxylistscom/issues',
            },
            repository => {
                type => 'git',
                url => 'https://github.com/stevieb9/p5-www-freeproxylistscom.git',
                web => 'https://github.com/stevieb9/p5-www-freeproxylistscom',
            },
        },
    },
    'PREREQ_PM' => {
               'Carp' => '1.04',
               'Class::Accessor::Grouped' => 0,
               'Devel::TakeHashArgs' => '0.001',
               'HTML::Entities' => '1.35',
               'HTML::TokeParser::Simple' => '3.15',
               'Test::More' => 0,
               'URI' => '1.35',
               'WWW::Mechanize' => '1.34'
             },
    'INSTALLDIRS' => 'site',
    'EXE_FILES' => [],
    'PL_FILES' => {}
);


lib/WWW/FreeProxyListsCom.pm
package WWW::FreeProxyListsCom;

use warnings;
use strict;

our $VERSION = '1.007';

use Carp;
use Data::Dumper;
use URI;
use WWW::Mechanize;
use HTML::TokeParser::Simple;
use HTML::Entities;
use Devel::TakeHashArgs;
use base 'Class::Accessor::Grouped';

__PACKAGE__->mk_group_accessors( simple => qw/
    error
    mech
    debug
    list
    filtered_list
    url_list
/);

sub new {
    my $self = bless {}, shift;

    get_args_as_hash(
        \@_, \my %args,
        {
            timeout => 30,
        }
    ) or croak $@;

    $args{mech} ||= WWW::Mechanize->new(
        timeout => $args{timeout},
        agent   => 'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.8.1.12)'
                    .' Gecko/20080207 Ubuntu/7.10 (gutsy) Firefox/2.0.0.12',
    );

    $self->mech( $args{mech} );
    $self->debug( $args{debug} );

    return $self;
}
sub get_list {
    my $self = shift;

    $self->$_(undef) for qw(error list);

    get_args_as_hash(\@_, \my %args, {
            type        => 'elite',
            max_pages   => 1,
        }
    ) or croak $@;

    my %page_for = (
        non_anonymous   => 'non-anonymous',
        map { $_ => $_ } qw(
                elite
                anonymous
                https
                standard
                socks
                us
                uk
                ca
                fr
        ),
    );

    exists $page_for{ $args{type} }
        or croak 'Invalid `type` argument was passed to fetch(). '
            . 'Must be one of' . join q|, |, keys %page_for;

    my $mech = $self->mech;
    my $page_type = $page_for{ $args{type} };
    my $url = $self->_url;

    my $uri = URI->new(
        "$url/$page_type.html"
    );

    $mech->get($uri)->is_success
        or return $self->_set_error($mech, 'net');

    $page_type eq 'anonymous'
        and $page_type = 'anon';
    $page_type eq 'non-anonymous'
        and $page_type = 'nonanon';

    # little tweaking to get the URI to the file normally loaded with AJAX
    my @links = map {
        "http://www.freeproxylists.com/load_${page_type}_" .
            ($_->url =~ m|([^/]+$)|)[0]
    } $mech->find_all_links(text_regex => qr/^detailed list #\d+/i);

    $args{max_pages}
        and @links = splice @links, 0, $args{max_pages};

    $self->debug
        and print "Going to fetch data from: \n" . join "\n", @links,'';

    my @proxies;
    for ( @links ) {
        unless ( $mech->get($_)->is_success ) {
            $self->debug
                and carp 'Network error: ' . $mech->res->status_line;
            next;
        }

        my $list_ref = $self->_parse_list( $mech->res->content )
            or next;

        push @proxies, @$list_ref;
    }

    return $self->list( \@proxies );
}
sub filter {
    my $self = shift;

    $self->$_(undef) for qw(error filtered_list);

    get_args_as_hash( \@_, \my %args)
        or croak $@;

    my %valid_filters;
    @valid_filters{ qw(ip  port  is_https  country  last_test  latency) }
    = (1) x 5;

    grep { not exists $valid_filters{$_} } keys %args
        and return $self->_set_error(
            'Invalid filter specified, valid ones are: '.
                join q|, |, keys %valid_filters
        );

    my $list_ref = $self->list
        or return $self->_set_error(
           'Proxy list seems to be undefined, did you call get_list() first?'
        );

    my @filtered;
    foreach my $proxy_ref ( @$list_ref ) {
        my $is_good = 0;
        for ( keys %args ) {
            if ( ref $args{$_} eq 'Regexp' ) {
                $proxy_ref->{$_} =~ /$args{$_}/
                    and $is_good++;
            }
            else {
                $proxy_ref->{$_} eq $args{$_}
                    and $is_good++;
            }
        }

        $is_good == keys %args
            and push @filtered, { %$proxy_ref };
    }
    return $self->filtered_list( \@filtered );
}
sub urls {
    my ($self) = @_;

    my $proxies = $self->filtered_list ? $self->filtered_list : $self->list;

    if (! @$proxies){
        $self->_set_error(
           'Proxy list seems to be undefined, did you call get_list() first?'
        );
    }

    my @urls;

    for (@$proxies){
        my $protocol = $_->{is_https} eq 'true'
            ? 'https'
            : 'http';

            push @urls, join '', "$protocol://", join ':', @$_{qw(ip port)};
    }

    return $self->url_list(\@urls);
}
sub _parse_list {
    my ( $self, $content ) = @_;

    # EVIL EVIL EVIL!! WEEE \o/
    ( $content ) = $content =~ m|<quote>(.+?)</quote>|s;
    decode_entities $content;

    my $parser = HTML::TokeParser::Simple->new( \$content );

    my %cells;
    @cells{ 1..6 } = qw(ip port is_https latency last_test country);
    my %nav;
    @nav{ qw(get_data level data_cell) } = (0) x 3;

    my @data;
    my %current;
    while ( my $t = $parser->get_token ) {
        if ( $t->is_start_tag('tr') ) {
            @nav{ qw(get_data level) } = (1, 1);
        }
        elsif ( $nav{get_data} == 1 and $t->is_start_tag('td') ) {
            $nav{level} = 2;
            $nav{data_cell}++;
        }
        elsif ( $nav{data_cell} and $t->is_text ) {
            $current{ $cells{ $nav{data_cell} } } = $t->as_is;
        }
        elsif ( $t->is_end_tag('tr') ) {
            @nav{ qw(level get_data data_cell) } = ( 3, 0, 0 );

            next unless keys %current;

            $current{ $_ } = 'N/A'
                for grep { !defined $current{$_} or !length $current{$_} }
                    values %cells;

            push @data, { %current };
            %current = ();
        }
    }

    shift @data; # quick and dirty fix to rid of bad data.
    return \@data;
}
sub _set_error {
    my ( $self, $mech_or_error, $type ) = @_;
    if ( defined $type and $type eq 'net' ) {
        $self->error('Network error: ' . $mech_or_error->res->status_line);
    }
    else {
        $self->error( $mech_or_error );
    }
    return;
}
sub _url {
    my ($self, $url) = @_;
    $self->{url} = $url if defined $url;
    return defined $self->{url} ? $self->{url} : 'http://freeproxylists.com';
}
1;
__END__

=encoding utf8

=head1 NAME

WWW::FreeProxyListsCom - Fetch proxy lists from http://www.freeproxylists.com

=for html
<a href="http://travis-ci.org/stevieb9/p5-www-freeproxylistscom"><img src="https://secure.travis-ci.org/stevieb9/p5-www-freeproxylistscom.png"/>
<a href='https://coveralls.io/github/stevieb9/p5-www-freeproxylistscom?branch=master'><img src='https://coveralls.io/repos/stevieb9/p5-www-freeproxylistscom/badge.svg?branch=master&service=github' alt='Coverage Status' /></a>


=head1 SYNOPSIS

    use strict;
    use warnings;

    use WWW::FreeProxyListsCom;

    my $fpl = WWW::FreeProxyListsCom->new;

    my $proxy_list = $fpl->get_list or die $prox->error;

    $proxy_list = $fpl->filter(type => 'is_https');

    my $formatted_into_urls = $fpl->urls;

=head1 DESCRIPTION

The module provides interface to fetch proxy server lists from
L<http://www.freeproxylists.com/>

=head1 CONSTRUCTOR

=head2 C<new>

    my $fpl = WWW::FreeProxyListCom->new;

    my $fpl2 = WWW::FreeProxyListCom->new(
        timeout => 20, # or 'mech'
        mech    => WWW::Mechanize->new(agent => 'foos', timeout => 20),
        debug   => 1
    );

Bakes up and returns a fresh WWW::FreeProxyListCom object. Takes a few
arguments, all of which are I<optional>. Possible arguments are as follows:

=head3 C<timeout>

    my $fpl = WWW::FreeProxyListCom->new(timeout => 10);

Takes a scalar as a value which is the value that will be passed to
the L<WWW::Mechanize> object to indicate connection timeout in seconds.
B<Defaults to:> C<30> seconds.

=head3 C<mech>

    my $fpl = WWW::FreeProxyListCom->new(
        mech => WWW::Mechanize->new(agent => '007', timeout => 10),
    );

If a simple timeout is not enough for your needs feel free to specify
the C<mech> argument which takes a L<WWW::Mechanize> object as a value.
B<Defaults to:> plain L<WWW::Mechanize> object with C<timeout> argument
set to whatever WWW::FreeProxyListCom's C<timeout> argument
is set to as well as C<agent> argument is set to mimic FireFox.

=head3 C<debug>

    my $fpl = WWW::FreeProxyListCom->new(debug => 1);

When set to a true value will make the object print out some debugging
info. B<Defaults to:> C<0>

=head1 METHODS

=head2 C<get_list>

    my $proxy_list = $fpl->get_list or die $fpl->error;

    my $fpl2 = $fpl->get_list(
        type      => 'standard',
        max_pages => 5,
    ) or die $fpl->error;

Instructs the object ot fetch a list of proxies from
L<http://www.freeproxylists.com/> website. On failure returns either
C<undef> or an empty list depending on the context and the reason
for failure will be available via C<error()> method. B<Note:> if request
for a each of the "list" (see C<max_pages> argument below) fails the
C<get_list()> will NOT error out, if you are getting empty proxy lists
try setting C<debug> option on in the constructor and it will carp()
any failures on the "list" gets. On success returns an arrayref of hashrefs,
see C<RETURN VALUE> section below for details. Takes several arguments all
of which are I<optional>. To understand them better you should visit
L<http://www.freeproxylists.com/> first. The possible arguments are
as follows:

=head3 C<type>

    type => $string

B<Optional>. Specifies the list of proxies to fetch. B<Defaults to:> C<elite>.
Possible arguments are as follows. Note all are plain HTTP except C<socks> and
C<https>.

    elite           = Elite (hides you entirely)
    anonymous       = Anonymous (hides you, but shows you're using a proxy)
    non_anonymous   = non-anonymous (no masking at all)
    https           = HTTPS (SSL enabled, may not hide you)
    standard        = standard HTTP/HTTPS/SOCKS/Proxy ports (may not hide you)
    ca              = Canada
    fr              = France
    us              = United States
    uk              = United Kingdom
    socks           = SOCKS (version 4/5)

=head3 C<max_pages>

    max_pages => $int

B<Optional>. Specifies how many "lists" to fetch. In other words, if
you go to list section titled "http elite proxies" you'll see several lists
in the table; the C<max_pages> specifies how many of those lists to fetch.
If C<max_pages> is larger than the number of available lists only the
number of available lists will be fetched. A special value of C<0> indicates
that the object should fetch all available lists for a specified C<type>.
B<Defaults to:> C<1> (which is more than enough).

=head3 RETURN VALUE

    $VAR1 = [
        {
            'country' => 'China',
            'last_test' => '3/15 4:23:14 pm',
            'ip' => '121.15.200.147',
            'latency' => '5115',
            'port' => '80',
            'is_https' => 'true'
        },
    ]

On success C<get_list()> method returns a (possibly empty) arrayref of
"proxy" hashrefs. The hashrefs represent each proxy listed on the proxy
list on the site. Each will contain the following keys (if the value for a
specific key was not found on the site it will be set to C<N/A>):

=over 10

=item ip

The IP address of the proxy

=item port

The port of the proxy

=item country

The country of the proxy

=item last_test

When was the proxy last tested to be alive, this is the "Date checked, UTC"
column on the site.

=item latency

Corresponds to the "Latency" column on the site

=item is_https

Corresponds to "HTTPS" column on the site.

=back

=head2 C<filter>

    my $filtered_list = $fpl->filter(
        port        => 80,
        ip          => qr/^120/,
        country     => 'Russia',
        is_https    => 'true',
        last_test   => qr|^3/15|, # march 15's
        latency     => qr/\d{1,2}/,
    );

Must be called after a successfull call to C<get_list()> will croak
otherwise. Takes one or more key/value pairs of arguments which specify
filtering rule. The keys are the same as the keys of "proxy" hashref
in the return value of the C<get_list()> method. Values can be either
simple scalars or regexes (C<qr//>). If value is a regex the corresponding
value in the "proxy" hashref will matched against the regex, otherwise
the C<eq> will be done. Returns an arrayref of "proxy" hashrefs in the
exact same format as C<get_list()> returns except filtered. In other words
calling C<< $prox->filter( port => 80, latency => qr/\d{1,2}/ ) >> will
return only proxies with port number C<80> and for which latency is a two
digit value. On failure returns either C<undef> or an empty list depending on
the context and the reason for the error will be available via C<error()>
method. Although, C<filter()> should not fail if you pass proper filter
arguments and call it after successfull C<get_list()>.

=head2 C<error>

    my $proxy_list = $fpl->get_list
        or die $fpl->error;

When either C<get_list()> or C<filter()> methods fail they will return
either C<undef> or an empty list depending on the context and the reason
for the failure will be available via C<error()> method. Takes no arguments,
returns a human parsable message explaining why C<get_list()> or C<filter()>
failed.

=head2 C<urls>

    my $proxy_urls = $fpl->urls;

Must be called after a successful call to C<get_list()>. Takes no arguments,
returns an array reference of proxies. Each element is a string, in the
following format: C<https://ip_address:port>.

=head2 C<mech>

    my $old_mech = $fpl->mech;

    $fpl->mech(WWW::Mechanize->new(agent => 'blah'));

Returns a L<WWW::Mechanize> object used for fetching proxy lists.
When called with an
optional argument (which must be a L<WWW::Mechanize> object) will use it
in any subsequent C<get_list()> calls.

=head2 C<list>

    my $cached_proxy_list = $fpl->list;

Must be called after a successfull call to C<get_list()>. Takes no arugments,
returns the same arrayref of hashrefs last call to C<get_list()> returned.

=head2 C<filtered_list>

    my $cached_filtered_proxy_list = $fpl->filtered_list;

Must be called after a successfull call to C<filter()>. Takes no arugments,
returns the same arrayref of hashrefs last call to C<filter()> returned.

=head2 C<debug>

    my $old_debug = $fpl->debug;

    $fpl->debug(1);

Returns a currently set debug flag (see C<debug> argument to constructor).
When called with an argument will set the debug flag to the value specified.

=head2 C<debug_list>

    my $cached_urls = $fpl->url_list;

Returns the same list after the last successful call to C<urls()> without the
module re-processing the proxies that were acquired in the previous call to
C<get_list()>.

=head1 AUTHOR

Zoffix Znet, C<< <zoffix at cpan.org> >>
(L<http://zoffix.com>, L<http://haslayout.net>)

Adopted on Feb 4, 2016 and currently maintained by:

Steve Bertrand C<< <steveb at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to L<https://github.com/stevieb9/p5-www-freeproxylistscom/issues>.

=head1 COPYRIGHT & LICENSE

Copyright 2016, 2018 Steve Bertrand

Copyright 2008 Zoffix Znet, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut
